### DISCUSSION TITLE
Example 8, Thermal conductivity calculations with the SSCHA crashed

### DISCUSSION THREAD

**Discussion by yusufshaidu** [SSCHAcode/python-sscha#186] [Q&A] [answered]
I am running example 6 tutorial the shows how to compute thermal conductivity with MLP but keep crashing at the first step.  

Here is the error message: 
"lattice = np.array(np.transpose(cell[0]), dtype="double", order="C")
TypeError: float() argument must be a string or a real number, not 'Atom'
"
The error seems to originate from "compute_phonons_finite_displacements_sym" called through compute_phonons_finite_displacements from the function "get_starting_dynamical_matrices".  Does anyone knows the source of such errors?

Thanks!
Yusuf

**Comment by mesonepigreco** [comment]
Hi Yusuf,
Can you past here the full error stack?

**Reply by yusufshaidu** [reply]
Here is the full output.

/pscratch/sd/y/yshaidu/Code/sscph_env/lib/python3.10/site-packages/sscha/aiida_ensemble.py:18: UserWarning: aiida or aiida-quantumespresso are not installed
  warnings.warn('aiida or aiida-quantumespresso are not installed')
                Step[ FC]     Time          Energy          fmax
*Force-consistent energies used in optimization.
BFGSLineSearch:    0[  0] 17:09:20       32.366892*     174.7496
BFGSLineSearch:    1[  1] 17:09:21       -7.695738*      10.9365
BFGSLineSearch:    2[  3] 17:09:21       -9.073873*       3.2856
BFGSLineSearch:    3[  4] 17:09:21       -9.246568*       0.8048
BFGSLineSearch:    4[  5] 17:09:21       -9.259009*       0.0943
BFGSLineSearch:    5[  6] 17:09:21       -9.259190*       0.0033
BFGSLineSearch:    6[  7] 17:09:21       -9.259190*       0.0000
Traceback (most recent call last):
  File "/pscratch/sd/y/yshaidu/lattice_thermal_conductivity/sscha/example_run1.py", line 49, in <module>
    dyn, bulk = get_starting_dynamical_matrices(
  File "/pscratch/sd/y/yshaidu/lattice_thermal_conductivity/sscha/example_run1.py", line 28, in get_starting_dynamical_matrices
    dyn = CC.Phonons.compute_phonons_finite_displacements(structure,
  File "/pscratch/sd/y/yshaidu/Code/sscph_env/lib/python3.10/site-packages/cellconstructor/Phonons.py", line 4498, in compute_phonons_finite_displacements
    return compute_phonons_finite_displacements_sym(structure, ase_calculator, epsilon,
  File "/pscratch/sd/y/yshaidu/Code/sscph_env/lib/python3.10/site-packages/cellconstructor/Phonons.py", line 4675, in compute_phonons_finite_displacements_sym
    symm = spglib.get_symmetry(super_structure.get_ase_atoms())
  File "/pscratch/sd/y/yshaidu/Code/sscph_env/lib/python3.10/site-packages/spglib/spglib.py", line 248, in get_symmetry
    lattice, _, _, magmoms = _expand_cell(cell)
  File "/pscratch/sd/y/yshaidu/Code/sscph_env/lib/python3.10/site-packages/spglib/spglib.py", line 1985, in _expand_cell
    lattice = np.array(np.transpose(cell[0]), dtype="double", order="C")
TypeError: float() argument must be a string or a real number, not 'Atom'

**Reply by diegomartinez2** [reply]
So this is the thermal conductivity example that takes the data from example 6. The error call is in spglib.py and comes from the structure data calculated in example 6 and read by ASE. This error may come from the original structure data if instead of the type of atom it says simply "Atom". 
Can you show us your '../06_the_SSCHA_with_machine_learning_potentials/POSCAR' file?

**Comment by mesonepigreco** [comment, marked as answer]
It seems the problem is much bigger.
Togo just removed support for ase atoms in spglib, thus breaking all the compatibility of our code with spglib.

https://spglib.readthedocs.io/en/latest/releases.html

For now, the solution is to downgrade spglib prior 2.3


```
pip remove spglib
pip install spglib<=2.2 --force-reinstall
```

**Reply by yusufshaidu** [reply]
Thanks you very much for your help. This problem has been solved! But I encountered another issues. 

/pscratch/sd/y/yshaidu/Code/sscph_env/lib/python3.10/site-packages/sscha/aiida_ensemble.py:18: UserWarning: aiida or aiida-quantumespresso are not installed
  warnings.warn('aiida or aiida-quantumespresso are not installed')
                Step[ FC]     Time          Energy          fmax
*Force-consistent energies used in optimization.
BFGSLineSearch:    0[  0] 12:05:24       32.366892*     174.7496
BFGSLineSearch:    1[  1] 12:05:24       -7.695738*      10.9365
BFGSLineSearch:    2[  3] 12:05:24       -9.073873*       3.2856
BFGSLineSearch:    3[  4] 12:05:24       -9.246568*       0.8048
BFGSLineSearch:    4[  5] 12:05:24       -9.259009*       0.0943
BFGSLineSearch:    5[  6] 12:05:24       -9.259190*       0.0033
BFGSLineSearch:    6[  7] 12:05:24       -9.259190*       0.0000
Number of symmetry inequivalent displacements: 1
Traceback (most recent call last):
  File "/pscratch/sd/y/yshaidu/lattice_thermal_conductivity/sscha/example_run1.py", line 66, in <module>
    relax.vc_relax(static_bulk_modulus = bulk,
  File "/pscratch/sd/y/yshaidu/Code/sscph_env/lib/python3.10/site-packages/sscha/Relax.py", line 604, in vc_relax
    self.minim.ensemble.compute_ensemble(
  File "/pscratch/sd/y/yshaidu/Code/sscph_env/lib/python3.10/site-packages/sscha/Ensemble.py", line 3848, in compute_ensemble
    computing_ensemble = self.get_noncomputed()
  File "/pscratch/sd/y/yshaidu/Code/sscph_env/lib/python3.10/site-packages/sscha/Ensemble.py", line 3989, in get_noncomputed
    return self.split(non_mask)
  File "/pscratch/sd/y/yshaidu/Code/sscph_env/lib/python3.10/site-packages/sscha/Ensemble.py", line 3942, in split
    ens.update_weights(self.current_dyn, self.current_T)
  File "/pscratch/sd/y/yshaidu/Code/sscph_env/lib/python3.10/site-packages/sscha/Ensemble.py", line 1950, in update_weights
    old_a = SCHAModules.thermodynamic.w_to_a(w, self.T0)
AttributeError: module 'SCHAModules' has no attribute 'thermodynamic'

I am not sure what I doing wrong because I see that there is a file in SCHAModules that contains thermodynamic.

**Reply by mesonepigreco** [reply]
Glad it solved the issue.
Regarding your second error, there is the solution in #184 
We are going to upgrade the code in the next days to fix these two problems

**Comment by yusufshaidu** [comment]
Thank you very much.  I look forward to the upgrade! 

In the meantime, I am now able to run the example but needed to change a little bit the new Thermodynamic python function. Basically, for instance, w_to_a needs not to take n as input as the shape of a could be infer from the shape of w (I think?) and also because n is not an argument of w_to_a in Ensemble.py: 

I replaced 
def w_to_a(w, T, n):
    a = np.zeros(n)
 with
 def w_to_a(w, T):
    a = np.zeros(w.shape)
and the other function that request n.

Please let me know if there is any problem with these changes. 

Many thanks.

**Comment by mesonepigreco** [comment]
No correct,
You are right, there is a mistake in that answer.
I will update the accepted answer

Il Sab 3 Feb 2024, 01:16 yusufshaidu ***@***.***> ha scritto:
